Suppress "unused variable" warning during including Boost.Flyweight library

# Details

Too many "unused variable" warnings are output to console in compiling due to `Boost.Flyweight` like this:

```
/path/to/boost/flyweight/detail/flyweight_core.hpp:204:22: warning: unused variable 'lock' [-Wunused-variable]
lock_type lock(mutex());
```

Add these lines to suppress the warnings:

```cpp
// For GCC and Clang
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wunused-variable"
// Include the headers here.
#pragma GCC diagnostic pop 

// For MSVC
// Currently the compiler does not report warnings about that.
#pragma warning(push)  
#pragma warning(disable: /*Warning ID*/)  
// Include the headers here.
#pragma warning(pop)   
```

cf. [Question about the way to disable GCC warnings temporarily in code(stack oveflow)](https://stackoverflow.com/questions/3378560/how-to-disable-gcc-warnings-for-a-few-lines-of-code)
cf. [Clang's manual](https://clang.llvm.org/docs/UsersManual.html#controlling-diagnostics-via-pragmas)

--- a/src/shared_id.hpp
+++ b/src/shared_id.hpp
@@ -1,8 +1,17 @@
 #pragma once
 #include <string>
+
+// Assumes that a compiler just ignores unknown pragmas.
+// Supresses "unused variable" warning due to Boost.Flyweight library.
+// This warning is reported only by Clang.
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunused-variable"
 #include <boost/flyweight.hpp>
 #include <boost/flyweight/no_locking.hpp>
 #include <boost/flyweight/no_tracking.hpp>
+#pragma clang diagnostic pop
+
+
 
 namespace elona
 {

